### PR TITLE
Add ping heartbeat command

### PR DESCRIPTION
**Pull request by QuentinBisson**
Towards https://github.com/giantswarm/giantswarm/issues/23238

### CODE REVIEW

File: cmd/ping.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+func runPing(opts *pingCmdOptions) {
+	repo, err := client.New(nil)
+	if err != nil {
+		log.Fatalf("Failed to init OpsGenie client: %v\n", err)
+	}
+	c := ctl.NewCtl(repo)
+
+	pings, err := c.Ping(opts.selectorOptions.ToConfig())
+	for hbn, ping := range pings {
```
Comment (mcbenjemaa) [inline cmd/ping.go:115]: Why you don't handle the error?

Comment (mcbenjemaa) [reply cmd/ping.go:115]: Actually, I think it's better to move this logic to the underlying function `c.Ping()` 
so that, we only have one single loop.

**Review by mcbenjemaa — CHANGES_REQUESTED** [review]

File: pkg/ctl/adapter.go [hunk 2]
```diff
@@ -78,6 +78,28 @@ func (c *ctl) Disable(opts *SelectorConfig) ([]heartbeat.HeartbeatInfo, error) {
...
+	}
+
+	heartbeats, err := c.Get(opts)
+	if err != nil {
+		return nil, err
+	}
+
+	var pingResults = make(map[string]heartbeat.PingResult)
+	for _, h := range heartbeats {
+		// TODO: context.TODO
```
[the remark this answers is not in the document]
Comment (QuentinBisson) [reply pkg/ctl/adapter.go]: Done

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,10 @@
+FROM alpine:3.16.2
+
+RUN apk update
+
+ENV USER root
```
[the remark this answers is not in the document]
Comment (QuentinBisson) [reply Dockerfile]: Fixed

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,10 @@
+FROM alpine:3.16.2
+
+RUN apk update
```
[the remark this answers is not in the document]
Comment (QuentinBisson) [reply Dockerfile]: Fixed

File: cmd/ping.go (hunk 1 above)
Comment (QuentinBisson) [reply cmd/ping.go]: The err was handled after hte loop.
I disagree with the loop management as logging to the console is a side effect of the function and I'm not sure we are really striving for that much optimization in that sample CLI :)

File: .circleci/config.yml [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+          filters:
+            branches:
+              only: main
+            # Needed to trigger job also on git tag.
+            tags:
+              only: /^v.*/
+
+      - architect/push-to-docker:
+          name: push-to-docker
+          image: quay.io/giantswarm/heartbeatctl
```
Comment (TheoBrigitte) [inline .circleci/config.yml:41]: What about other registries?

File: .nancy-ignore [hunk 6]
```diff
@@ -0,0 +1,2 @@
+# pkg:golang/github.com/gogo/protobuf@v1.3.1
+CVE-2021-3121  until=2023-01-01
```
Comment (TheoBrigitte) [inline .nancy-ignore:2]: Sounds like 1.3.2 fixes this https://nvd.nist.gov/vuln/detail/cve-2021-3121
```
An issue was discovered in GoGo Protobuf before 1.3.2. plugin/unmarshal/unmarshal.go lacks certain index validation, aka the "skippy peanut butter" issue.
```

Please upgrade the dependency https://github.com/gogo/protobuf/releases/tag/v1.3.2

File: Dockerfile [hunk 7]
```diff
@@ -0,0 +1,6 @@
+FROM alpine:3.16.2
+
+ENV PATH /usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
```
Comment (TheoBrigitte) [inline Dockerfile:3]: Why redefining the default PATH?

```
$ docker run --rm -it alpine:3.16.2 ash -c 'echo $PATH'
/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
```

I would suggest to remove this
```suggestion
```

File: cmd/ping.go [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+)
+
+// pingCmdOptions holds values for options accepted by the ping command
+type pingCmdOptions struct {
+	selectorOptions *cmdutil.SelectorOptions
+}
+
+var (
+	pingDocLong = heredoc.Doc(`
+		Ping specified heartbeats to remove their expiration status.
```
Comment (TheoBrigitte) [inline cmd/ping.go:22]: We are not removing anything here, this feels wrong. I would prefer something like this :

```suggestion
		Issue a ping request to specified heartbeats.
```

File: cmd/ping.go [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+
```
Comment (TheoBrigitte) [inline cmd/ping.go:23]: Also something important from opsgenie doc and probably worth mentioning here.

`receiving a successful response does not necessarily mean that the heartbeat exists`

from: https://docs.opsgenie.com/docs/heartbeat-api#ping-heartbeat-request

**Review by TheoBrigitte — CHANGES_REQUESTED** [review]
Overall looks good.

But I would appreciate when this PR is split into mutliple PRs:

- update the circleci config
- ping command boilerplate
- ping command implementation/logic
- refactoring/renaming

Also showing an example run would be nice.

File: .nancy-ignore (hunk 6 above)
Comment (QuentinBisson) [reply .nancy-ignore:2]: Oh I misread it was still in 1.3.2 and that 1.3.2 was the last version available, let's try

File: Dockerfile [hunk 10]
```diff
@@ -0,0 +1,6 @@
...
+
+ADD ./heartbeatctl /usr/local/bin/heartbeatctl
+ENTRYPOINT ["/usr/local/bin/heartbeatctl"]
```
[the remark this answers is not in the document]
Comment (QuentinBisson) [reply Dockerfile:6]: I did not do it on purpose :)

File: cmd/ping.go (hunk 8 above)
Comment (QuentinBisson) [reply cmd/ping.go:22]: Fixed in https://github.com/giantswarm/heartbeatctl/pull/15

File: cmd/ping.go [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+		are '=', '==', '!='.
+
+		And finally any positional arguments are taken as regular expressions to match
+		against heartbeat names. Multiple arguments can be given and they will be joined
+		into an or-expression and wrapped in beginning and end-of-string bounds so the
+		expressions have to match the entire name. E.g. parameters 'foo' 'bar-.*' will
+		result in a regex '^(foo|bar-.*)$'.
+	`)
+	pingDocExamples = heredoc.Doc(`
+		# ping all heartbeats with specified label 'managed-by' equal to 'foobricator'
```
[the remark this answers is not in the document]
Comment (QuentinBisson) [reply cmd/ping.go:48]: Done in https://github.com/giantswarm/heartbeatctl/pull/15

**Pull request closed without merging**